ddprof-lib/src/main/cpp/wallClock.cpp

```diff
@@ -161,49 +161,83 @@ void WallClockJVMTI::timerLoop() {
-        jvmtiEnv* jvmti = VM::jvmti();
-        if (jvmti == nullptr) {
-            return;
-        }
-        JNIEnv* jni = VM::jni();
-
-        jint threads_count = 0;
-        jthread* threads_ptr = nullptr;
-        jvmti->GetAllThreads(&threads_count, &threads_ptr);
+      jvmtiEnv* jvmti = VM::jvmti();
+      if (jvmti == nullptr) {
+          return;
+      }
+      JNIEnv* jni = VM::jni();
 
-        bool do_filter = Profiler::instance()->threadFilter()->enabled();
-        int self = OS::threadId();
+      jint threads_count = 0;
+      jthread* threads_ptr = nullptr;
+      jvmti->GetAllThreads(&threads_count, &threads_ptr);
 
-        for (int i = 0; i < threads_count; i++) {
-          jthread thread = threads_ptr[i];
-          if (thread != nullptr) {
-            VMThread* nThread = VMThread::fromJavaThread(jni, thread);
-            if (nThread == nullptr) {
-              continue;
-            }
-            int tid = nThread->osThreadId();
-            if (tid != self && (!do_filter || Profiler::instance()->threadFilter()->accept(tid))) {
-              threads.push_back({nThread, thread});
-            }
+      bool do_filter = Profiler::instance()->threadFilter()->enabled();
+      int self = OS::threadId();
+      for (int i = 0; i < threads_count; i++) {
+        jthread thread = threads_ptr[i];
+        if (thread != nullptr) {
+          VMThread* nThread = VMThread::fromJavaThread(jni, thread);
+          if (nThread == nullptr) {
+            jni->DeleteLocalRef(thread);
+            continue;
           }
+          jint thread_state;
+          if (jvmti->GetThreadState(thread, &thread_state) == JVMTI_ERROR_NONE &&
+              (thread_state & JVMTI_THREAD_STATE_TERMINATED) == 0) {
+              int tid = VMThread::nativeThreadId(jni, thread);
+              if (tid != -1 && tid != self && (!do_filter || Profiler::instance()->threadFilter()->accept(tid))) {
+                  threads.push_back({nThread, jni->NewWeakGlobalRef(thread)});
+              }
+          }
+          jni->DeleteLocalRef(thread);
         }
-        jvmti->Deallocate((unsigned char*)threads_ptr);
+      }
+      jvmti->Deallocate((unsigned char*)threads_ptr);
     };
 
   auto sampleThreads = [&](ThreadEntry& thread_entry, int& num_failures, int& threads_already_exited, int& permission_denied) {
     static jint max_stack_depth = (jint)Profiler::instance()->max_stack_depth();
     static jvmtiFrameInfo* frame_buffer = new jvmtiFrameInfo[max_stack_depth];
     static jvmtiEnv* jvmti = VM::jvmti();
+    static JNIEnv* jni = VM::jni();
 
     int num_frames = 0;
-    jvmtiError err = jvmti->GetStackTrace(thread_entry.java, 0, max_stack_depth, frame_buffer, &num_frames);
+    if (thread_entry.java_ref == nullptr) {
+      num_failures++;
+      return false;
+    }
+
+    jobject thread = jni->NewLocalRef(thread_entry.java_ref);
+    if (thread == nullptr) {
+        num_failures++;
+        jni->DeleteWeakGlobalRef(thread_entry.java_ref);
```

Comment:
this can probably be easier to manage with some RAII design (like defer deletion when something goes out of scope)

Reply:
I have added RAII objects to make this easier to grasp